odl/solvers/functional/functional.py

```diff
@@ -0,0 +1,1671 @@
+        of ``f``, this corresponds to the functional ``f(. - y)``.
+
+        Parameters
+        ----------
+        func : `Functional`
+            Functional which is to be translated.
+        translation : `LinearSpaceVector`
+            Element in ``func.domain``, with which the argument is translated.
+        """
+
+        if not isinstance(func, Functional):
+            raise TypeError('`func` {!r} not a `Functional` instance'
+                            ''.format(func))
+
+        if translation not in func.domain:
+            raise TypeError('`translation` {!r} not in func.domain {!r}'
+                            ''.format(translation.space, func.domain))
+
+        super().__init__(domain=func.domain, linear=False,
+                         grad_lipschitz=func.grad_lipschitz)
+
+    # TODO: Add case if we have translation -> scaling -> translation?
+        if isinstance(func, TranslatedFunctional):
+            self._original_func = func.original_func
+            self._translation = func.translation + translation
+
+        else:
+            self._original_func = func
+            self._translation = translation
+
+    @property
+    def original_func(self):
+        return self._original_func
+
+    @property
+    def translation(self):
+        return self._translation
+
+    def _call(self, x):
+        """Evaluates the functional in a point ``x``.
+
+        Parameters
+        ----------
+        x : `LinearSpaceVector`
+            Element in the domain of the functional.
+            The point in which the functional is evaluated
+
+        Returns
+        -------
+        `self(x)` : `float`
+            Evaluation of the functional, which is a constant.
+        """
+        return self._original_func(x - self._translation)
+
+    @property
+    def gradient(self):
+        """Gradient operator of the functional.
+
+        The operator is given by a transation of the gradient operator of the
+        original functional.
+
+        Notes
+        -----
+        The operator that corresponds to the mapping
+
+        .. math::
+
+            x \\to \\nabla f(x)
+
+        where :math:`\\nabla f(x)` is the element used to evaluate
+        derivatives in a direction :math:`d` by
+        :math:`\\langle \\nabla f(x), d \\rangle`.
+        """
+
+        class TranslatedGradientOperator(Operator):
+
+            """The gradient operator for a translated functional."""
+
+            def __init__(self, translated_func, translation):
+                """Initialize a new instance.
+
+                Parameters
+                ----------
+                translated_func : `Functional`
+                    Functional corresponding to ``F(. - translation)``
+                translation : `LinearSpaceVector`
+                    The translation with which ``translated_func`` has been
+                    translated.
+
+                """
+                # TODO: is there a good way to see if this operator is linear?
```

Comment:
Dunno. But in 95 % of the interesting cases, the gradient operator is not linear, so if we go for `linear=False`, it should be okay.